From 0ab3397479c688b26e8c39b90b090e53ad8b01a9 Mon Sep 17 00:00:00 2001
From: Josef Holzmayr <jester@theyoctojester.info>
Date: Wed, 8 Apr 2026 19:28:00 +0000
Subject: [PATCH] configs: efi-x86_app64: enable mender requirements

Configure efi-x86_app64_defconfig for Mender A/B boot support:
- Redundant env stored as FAT files on partition 2 (dedicated env
  partition, separate from ESP) to prevent OVMF boot discovery
  failures caused by FAT writes to the ESP.
- Bootcount limit for automatic rollback on failed updates.
- Mender-aware BOOTCOMMAND with conditional kernel load (&&) to
  prevent booting stale kernel data from RAM on load failure.
- Load address 0x70000000 (above EFI app reserved memory region).
- LMB_ARCH_MEM_MAP disabled to allow memory access in EFI app mode.

Upstream-Status: Inappropriate [Mender specific]
Signed-off-by: Josef Holzmayr <jester@theyoctojester.info>
---
 configs/efi-x86_app64_defconfig | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/configs/efi-x86_app64_defconfig b/configs/efi-x86_app64_defconfig
index e0cfe3e..31c3602 100644
--- a/configs/efi-x86_app64_defconfig
+++ b/configs/efi-x86_app64_defconfig
@@ -1,6 +1,6 @@
 CONFIG_X86=y
 CONFIG_NR_DRAM_BANKS=8
-CONFIG_ENV_SIZE=0x1000
+CONFIG_ENV_SIZE=0x20000
 CONFIG_DEFAULT_DEVICE_TREE="efi-x86_app"
 CONFIG_DEBUG_UART_BASE=0x0
 CONFIG_DEBUG_UART_CLOCK=0
@@ -14,7 +14,7 @@ CONFIG_SHOW_BOOT_PROGRESS=y
 CONFIG_USE_BOOTARGS=y
 CONFIG_BOOTARGS="root=/dev/sdb3 init=/sbin/init rootwait ro"
 CONFIG_USE_BOOTCOMMAND=y
-CONFIG_BOOTCOMMAND="ext2load scsi 0:3 01000000 /boot/vmlinuz; zboot 01000000"
+CONFIG_BOOTCOMMAND="run mender_setup; setenv bootargs console=ttyS0,115200 root=\${mender_kernel_root} rootwait rw; load \${mender_uboot_root} 0x70000000 /boot/bzImage && zboot 0x70000000 \$filesize; run mender_try_to_recover"
 CONFIG_SYS_CONSOLE_INFO_QUIET=y
 CONFIG_DISPLAY_BOARDINFO_LATE=y
 CONFIG_HUSH_PARSER=y
@@ -46,3 +46,13 @@ CONFIG_CMD_DHRYSTONE=y
 # CONFIG_GZIP is not set
 CONFIG_EFI=y
 CONFIG_EFI_APP_64BIT=y
+CONFIG_BOOTCOUNT_ENV=y
+CONFIG_BOOTCOUNT_LIMIT=y
+CONFIG_ENV_IS_IN_FAT=y
+CONFIG_ENV_FAT_INTERFACE="efi"
+CONFIG_ENV_FAT_DEVICE_AND_PART="0:2"
+CONFIG_ENV_FAT_FILE="uboot.env"
+CONFIG_SYS_REDUNDAND_ENVIRONMENT=y
+CONFIG_ENV_FAT_FILE_REDUND="uboot-redund.env"
+CONFIG_FAT_WRITE=y
+CONFIG_LMB_ARCH_MEM_MAP=n
-- 
2.34.1

